x/concentrated-liquidity/internal/math/tick.go

```diff
@@ -1,52 +1,117 @@
+func TickToPrice(tickIndex, kAtPriceOne sdk.Int) (price sdk.Dec, err error) {
+	if tickIndex.IsZero() {
+		return sdk.OneDec(), nil
 	}
-	if err != nil {
-		return sdk.Dec{}, err
+
+	if kAtPriceOne.LT(types.PrecisionValueAtPriceOneMin) || kAtPriceOne.GT(types.PrecisionValueAtPriceOneMax) {
+		return sdk.Dec{}, fmt.Errorf("kAtPriceOne must be in the range (%s, %s)", types.PrecisionValueAtPriceOneMin, types.PrecisionValueAtPriceOneMax)
 	}
 
-	sqrtPrice = sqrtPriceOsmoMath.SDKDec()
+	// The formula is as follows: k_increment_distance = 9 * 10**(-k_at_price_1)
+	// Due to sdk.Power restrictions, if the resulting power is negative, we take 9 * (1/10**k_at_price_1)
+	kIncrementDistance := sdkNineDec.Mul(handleNegativeExponents(kAtPriceOne.Neg()))
+
+	// Use floor division to determine how many k increments we have passed
+	kDelta := tickIndex.ToDec().Quo(kIncrementDistance).TruncateInt()
+
+	// Calculate the current k value from the starting k value and the k delta
+	curK := kAtPriceOne.Add(kDelta)
+
+	// Knowing what our curK is, we can then figure out what power of 10 this k corresponds to
+	curIncrement := handleNegativeExponents(curK)
 
-	return sqrtPrice, nil
+	// Now, starting at the minimum tick of the current increment, we calculate how many ticks in the current k we have passed
+	numAdditiveTicks := tickIndex.ToDec().Sub(kDelta.ToDec().Mul(kIncrementDistance))
+
+	// Finally, we can calculate the price
+	price = handleNegativeExponents(kDelta).Add(numAdditiveTicks.Mul(curIncrement))
+
+	return price, nil
 }
 
 // PriceToTick takes a price and returns the corresponding tick index
-func PriceToTick(price sdk.Dec) sdk.Int {
-	tick := osmomath.BigDecFromSDKDec(price).CustomBaseLog(TickBase)
-	return tick.SDKDec().TruncateInt()
+func PriceToTick(price sdk.Dec, kAtPriceOne sdk.Int) (sdk.Int, error) {
+	if price.Equal(sdk.OneDec()) {
+		return sdk.ZeroInt(), nil
+	}
+
+	if kAtPriceOne.LT(types.PrecisionValueAtPriceOneMin) || kAtPriceOne.GT(types.PrecisionValueAtPriceOneMax) {
+		return sdk.Int{}, fmt.Errorf("kAtPriceOne must be in the range (%s, %s)", types.PrecisionValueAtPriceOneMin, types.PrecisionValueAtPriceOneMax)
+	}
+
+	// The formula is as follows: k_increment_distance = 9 * 10**(-k_at_price_1)
+	// Due to sdk.Power restrictions, if the resulting power is negative, we take 9 * (1/10**k_at_price_1)
+	kIncrementDistance := sdkNineDec.Mul(handleNegativeExponents(kAtPriceOne.Neg()))
+
+	// Initialize the total price to 1, the current k to k_at_price_1, and the number of ticks passed to 0
+	total := sdk.OneDec()
+	ticksPassed := sdk.ZeroInt()
+	currentK := kAtPriceOne
+
+	// Set the currentIncrement to the kAtPriceOne
+	curIncrement := handleNegativeExponents(currentK)
+
+	// Now, we loop through the k increments until we have passed the price
+	// Once we pass the price, we can determine what which k values we have filled in their entirety,
+	// as well as how many ticks that corresponds to
+	for total.LT(price) {
```

Comment:
I have no solid understanding of the time complexity here. This function seems to be called every swap step. Each swap step, we start searching from the price of 1 until the actual price. This seems like O(k * n) time complexity where k is the number of swap steps and n is the price difference from 1.

I think this could be optimized. For every new swap step we are unlikely to drift too far away from the previous price. As a result, we can start searching from the old price, not from 1. 

I'm not sure how worthwhile it is of an improvement. But the potential O(k * n) seems concerning. Would like to know what other reviewers think

Reply:
another idea is to binary search this

Reply:
Another idea is to take a log base 10 of price via osmomath and truncate it. This will give us a geometric part of the ticks or kDelta. Then, we can binary search the additive range (maybe there is even an easier way to compute it)

Again, not sure if I'm overthinking the performance implications here

Reply:
We should only loop through this a maximum of the number of k values there are. So for instance, with a maximum of a k value of -12 and min of -1, we should only loop through this 11 times.

Agreed there is probably a better way to do this though, down to try these suggestions. Do you have any good recommendation on benchmarking?

Reply:
> Do you have any good recommendation on benchmarking?

I think we can try creating a benchmark around `calcOutAmtGivenIn`, initialize a pool close to max spot price, initialize many positions and try executing a large swap where kAtPriceOne is the smallest one possible. I think this would help us benchmark the worst-case scenario. Such a benchmark would be useful even outside of the context of `PriceToTick`

I don't think it's worth blocking this PR on this at this moment. If you don't a  relatively quick and trivial way to simplify this, we might want to capture this thread in an issue and punt it until later